src/types/connection/mod.rs

```diff
@@ -10,29 +10,98 @@ pub use connection_type::Connection;
+        after: Cursor,
+    },
+    /// Return all results before the cursor
+    Before {
+        /// Before this cursor
+        before: Cursor,
+    },
+    /// Return all results between the cursors
+    Between {
+        /// After this cursor
+        after: Cursor,
+        /// But before this cursor
+        before: Cursor,
+    },
+    /// Return the amount of results specified by `limit`, starting from the beginning
+    First {
+        /// The maximum amount of results to return
+        limit: usize,
+    },
+    /// Return the amount of results specified by `limit`, starting after the cursor
+    FirstAfter {
+        /// The maximum amount of results to return
         limit: usize,
+        /// After this cursor
+        after: Cursor,
     },
-    /// Backward query
-    Backward {
+    /// Return the amount of results specified by `limit`, starting from the beginning but ending before the cursor
+    FirstBefore {
+        /// The maximum amount of results to return
+        limit: usize,
         /// Before this cursor
-        before: Option<&'a str>,
-
-        /// How many records did this query return
+        before: Cursor,
+    },
+    /// Return the amount of results specified by `limit`, but between the cursors. Limit includes beginning results.
+    FirstBetween {
+        /// The maximum amount of results to return
         limit: usize,
+        /// After this cursor
+        after: Cursor,
+        /// But before this cursor
+        before: Cursor,
     },
+    /// Return the amount of results specified by `limit`, but before the end
+    Last {
+        /// The maximum amount of results to return
+        limit: usize,
+    },
+    /// Return the amount of results specified by `limit`, but before the end. Must not include anything before the cursor.
+    LastAfter {
+        /// The maximum amount of results to return
+        limit: usize,
+        /// After this cursor
+        after: Cursor,
+    },
+    /// Return the amount of results specified by `limit`, but before the cursor
+    LastBefore {
+        /// The maximum amount of results to return
+        limit: usize,
+        /// Before this cursor
+        before: Cursor,
+    },
+    /// Return the amount of results specified by `limit`, but between the cursors. Limit includes ending results.
+    LastBetween {
+        /// The maximum amount of results to return
+        limit: usize,
+        /// After this cursor
+        after: Cursor,
+        /// But before this cursor
+        before: Cursor,
+    },
+    /// An invalid query was made. For example: sending `first` and `last` in the same query
+    Invalid,
 }
 
 /// Empty edge extension object
 #[async_graphql_derive::SimpleObject(internal)]
 pub struct EmptyEdgeFields;
 
+// Temporary struct for to store values for pattern matching
+struct Pagination {
+    after: Option<Cursor>,
+    before: Option<Cursor>,
+    first: Option<i32>,
+    last: Option<i32>,
+}
```

Comment:
Having this implementing `InputObject` could be useful, so we can also use it directly in queries, ie:
```gql
query {
  users(page: { first: 10 }) { ... }
}
```

Reply:
Technically that's against the Relay spec, but we could move it outside the connection namespace.

Reply:
In practice I find it convenient so I can separate pagination from filters in arguments, but that's a matter of taste I guess. You can ignore my previous comment, anyone can still go with this pattern if they want in their own project.

Reply:
> This is what I was about to go with:
> 
> ```rust
> use async_graphql::{Cursor, FieldError, FieldResult};
> use std::convert::TryFrom;
> use std::ops::Range;
> 
> #[async_graphql::InputObject]
> pub struct ConnectionArgs {
>     #[field]
>     pub first: Option<i32>,
>     #[field]
>     pub after: Option<Cursor>,
>     #[field]
>     pub last: Option<i32>,
>     #[field]
>     pub before: Option<Cursor>,
> }
> 
> impl ConnectionArgs {
>     pub fn get_range<E, T>(&self, nodes: &[(Cursor, E, T)]) -> FieldResult<Range<usize>> {
>         let first = self.first.map(usize::try_from).map(Result::ok).flatten();
>         let last = self.last.map(usize::try_from).map(Result::ok).flatten();
>         let len = nodes.len();
> 
>         match (self.after.as_ref(), self.before.as_ref()) {
>             (Some(after), None) => {
>                 let start = nodes
>                     .iter()
>                     .position(|(cursor, _, _)| *cursor == *after)
>                     .map(|i| i + 1)
>                     .unwrap_or(0);
>                 let end = (start + first.unwrap_or(len)).min(nodes.len());
> 
>                 Ok(start..end)
>             }
>             (None, Some(before)) => {
>                 let end = nodes
>                     .iter()
>                     .position(|(cursor, _, _)| *cursor == *before)
>                     .unwrap_or(len);
>                 let start = end.saturating_sub(last.unwrap_or(len));
> 
>                 Ok(start..end)
>             }
>             (None, None) => match (first, last) {
>                 (Some(first), None) => Ok(0..first.min(len)),
>                 (None, Some(last)) => Ok(len.saturating_sub(last)..len),
>                 (None, None) => Ok(0..len),
> 
>                 (Some(_), Some(_)) => Err(FieldError::from(
>                     "[ConnectionArgs] first and last are mutually exclusive.",
>                 )),
>             },
>             (Some(_), Some(_)) => Err(FieldError::from(
>                 "[ConnectionArgs] before and after are mutually exclusive.",
>             )),
>         }
>     }
> }
> ```

@IcanDivideBy0 Now all the #[field] can be omitted, so you can define inputobject like this.

```rust
#[async_graphql::InputObject]
pub struct ConnectionArgs {
    pub first: Option<i32>,
    pub after: Option<Cursor>,
    pub last: Option<i32>,
    pub before: Option<Cursor>,
}
```